### PR TITLE
`SandboxFolder`: remove option to create in profile repository

### LINKED ISSUE #5494: Do not use `Profile.repository_uri` as the location for sandbox creation
The `repository_uri` key in a profile configuration is used for two purposes. It is the base path for both the repository container, but also for the directory where `SandBoxFolder`s are created. Really this key should only be used for the location of the repository. The location of the sandbox folders should simply be in the tmp folder of the operating system. This decoupling will make it possible to introduce storage backends in the future that do not have a path on the local file system for a file repository.

### PR DESCRIPTION
**Pull request by sphuber**
Fixes #5494 

The `SandboxFolder` by default would create the temporary directory in
the `sandbox` subfolder of the repository directory, defined by the
`repository_uri` key in the profile configuration.

However, this folder path should mainly serve to designate the location
of the file repository of the storage backend. So far the sandbox folder
hijacking the same key wasn't a problem as there was only one storage
backend implementation. However, when we allow other storage backend
implementations, having to keep this config key just for the sandbox
folder doesn't make sense. It would be more logical to always create the
sandbox folders in the temporary directory created through the builtin
method `tempfile.mkdtemp`.

### CODE REVIEW

**Comment by sphuber** [comment]
@giovannipizzi do you maybe still know what the original reasoning was to having the sandbox folder in a subpath of `repository_uri` as opposed in the temp directory of the OS? Was this out of concern for guaranteeing writing rights? Or maybe for backup reasons? I don't see why you would want to backup the contents of the sandbox folders though. I now actually have often lots of old folders and files in the sandbox that I have to clean manually. Would be better I think to use `/tmp` so this is cleaned more automatically.

**Comment by giovannipizzi** [comment]
In general you want to ensure the sandbox is on the same filesystem (and /tmp is often in another one), as this can guarantee atomic move of the files once you want to transfer them from sandbox to actual repository. This was needed before, and is needed inside the disk-objectstore. For the sandbox of AiiDA, (possibly?) this might not be needed anymore (but I'm not sure). Note also that we need to check with method we use to move files from sandbox to repository - some that guarantee atomicity don't work (=fail) between different filesystems (I think it's os.rename, but I'd need to double check).

Even if the limitation is not there, also note that people might put the repository on a large filesystem, while TMP might be on a different one (even in RAM, or a smaller FS). So I could imagine some rare but possible situations (I actually had one when trying to migrate an archive - I had to set an ENV VAR to change the location of the TMP folder) where there's not enough space on the TMP partition and the process fails (in this case adding large files to a node), even if there is a lot of empty space in the partition of the repository. So I'm a bit hesitant - maybe we should have an optional key for the tmp location, and by default if missing we keep the old (current) behaviour?

**Comment by sphuber** [comment]
> For the sandbox of AiiDA, (possibly?) this might not be needed anymore (but I'm not sure).

This is no longer needed, because we no longer use an atomic move. The reason is that we no longer assume that the file repository interface can use move operations. The files are therefore always copied through streams.

> some that guarantee atomicity don't work (=fail) between different filesystems (I think it's os.rename, but I'd need to double check).

Indeed it is `os.rename` that requires it to be on the same filesystem. Across filesystems one should use `shutil.move` instead. But as mentioned above, this is no longer applicable.

> So I'm a bit hesitant - maybe we should have an optional key for the tmp location, and by default if missing we keep the old (current) behaviour?

It is ok to keep it I guess, I just think it shouldn't be hijacking the `repository_uri` key from the profile. We should just make it a default path in the profile folder. It can simply be statically defined as `Config.dirpath / Profile.name / 'sandbox'`.

**Comment by chrisjsewell** [comment]
>  I just think it shouldn't be hijacking the `repository_uri` key from the profile.

I think you mean, you agree what I'd already said should happen in #5330 😉 

>  It can simply be statically defined as `Config.dirpath / Profile.name / 'sandbox'`.

No, it should be a configuration option on the profile, with a default, not hard-coded

**Comment by sphuber** [comment]
> No, it should be a configuration option on the profile, with a default, not hard-coded

Fine by me. It was never configurable before, but I have added it.

**Comment by chrisjsewell** [comment]
> Fine by me. It was never configurable before, but I have added it.

Cool cheers, the point is that it was always in an "accessible" place before, where you could inspect it if necessary, but now it will not be.

> where there's not enough space on the TMP partition and the process fails

I'd note also, "cleaning" of sandbox folders is not particularly reliable, since it's only deleted on (python) garbage collection. So if the process doesn't stop cleanly it won't be removed.
I guess this is improved by having it in a tmp folder, since at least it will always automatically be removed when restarting your system

**Comment by chrisjsewell** [comment]
Note, the goal (in this PR or another) should be to fully remove from `Profile`: https://github.com/aiidateam/aiida-core/blob/80d11258075c1a251c7d2bd348e46957d4766267/aiida/manage/configuration/profile.py#L219

and place it on the `PsqlDosBackend`, since it is purely specific to the disk-objectstore

**Comment by sphuber** [comment]
> I guess this is improved by having it in a tmp folder, since at least it will always automatically be removed when restarting your system

This was also my argument indeed to use that as a default. I personally have a lot of junk in the `repository/sandbox` folders, and I doubt that a lot of users even know that it exists and look there.

> Note, the goal (in this PR or another) should be to fully remove from Profile

Agree, happy to create a follow-up PR to do this.

@chrisjsewell can you also review the code changes perhaps?

**Comment by giovannipizzi** [comment]
For the junk in the sandbox - probably we should add the cleaning of the sandbox to the `verdi storage maintain` - with the full option, we already know the profile is locked so we can clean up safely that folder. If you agree, OK to do in a different PR for me (then, maybe open an issue if you agree?)

**Comment by sphuber** [comment]
@giovannipizzi  See https://github.com/aiidateam/aiida-core/issues/5514

**Comment by sphuber** [comment]
I think we should move ahead with this and leave the advanced functionality for #5514 . These changes are not making things worse. The sandbox folder is currently also not cleaned automatically or performed by some CLI command

File: aiida/cmdline/commands/cmd_archive.py [hunk 1]
```diff
@@ -435,8 +437,9 @@ def _import_archive_and_migrate(archive: str, web_based: bool, import_kwargs: di
     from aiida.tools.archive.imports import import_archive as _import_archive
 
     archive_format = get_format()
+    filepath = ctx.obj['config'].get_option('storage.sandbox') or None
```
Comment (ramirezfranciscof) [inline aiida/cmdline/commands/cmd_archive.py:440]: Why do we have/need two different ways of loading this? I see below for the daemon workers you use:

```python
from aiida.manage.configuration import get_config_option

filepath_sandbox = get_config_option('storage.sandbox') or None
```

File: aiida/engine/processes/calcjobs/calcjob.py [hunk 2]
```diff
@@ -429,11 +429,14 @@ def _perform_import(self):
         from aiida.common.datastructures import CalcJobState
         from aiida.common.folders import SandboxFolder
         from aiida.engine.daemon.execmanager import retrieve_calculation
+        from aiida.manage import get_config_option
```
Comment (ramirezfranciscof) [inline aiida/engine/processes/calcjobs/calcjob.py:432]: A bit of a nitpick, but do we have any guidelines as to how to import features that are exposed at higher levels? Here you are importing `get_config_option` from `aiida.manage`, but previously you did it from the more specific `aiida.manage.configuration`.

File: aiida/storage/sqlite_temp/backend.py [hunk 3]
```diff
@@ -133,7 +133,7 @@ def get_repository(self) -> SandboxRepositoryBackend:
             raise ClosedStorage(str(self))
         if self._repo is None:
             # to-do this does not seem to be removing the folder on garbage collection?
-            self._repo = SandboxRepositoryBackend(sandbox_in_repo=True)
+            self._repo = SandboxRepositoryBackend()
```
Comment (ramirezfranciscof) [inline aiida/storage/sqlite_temp/backend.py:136]: In all the other replacements I've seen you first do:

```python
filepath_sandbox = get_config_option('storage.sandbox') or None
```

And then provide that when creating the `SandboxRepositoryBackend` or the `SandboxFolder`, yet here it will always create the temporary folder using the temp location of the system. Was this intended? Not saying is wrong, just checking the difference, specially since here it is now changing the behavior.

File: tests/cmdline/commands/test_archive_import.py [hunk 4]
```diff
@@ -7,9 +7,9 @@
 # For further information on the license, see the LICENSE.txt file        #
 # For further information please visit http://www.aiida.net               #
 ###########################################################################
-"""Tests for `verdi import`."""
+# pylint: disable=redefined-outer-name
+"""Tests for `verdi archive import`."""
 from click.exceptions import BadParameter
-from click.testing import CliRunner
 import pytest
```
Comment (ramirezfranciscof) [inline tests/cmdline/commands/test_archive_import.py:13]: Referring to the whole file, not just these imports: Oh wow, github really goes haywire when you change a couple of indentations 😅. Why did you change this here? I don't see any modifications related to the sandbox folder feature, just the change from a class with an ``init`` to methods using the `run_cli_command`, am I missing anything?

**Review by ramirezfranciscof — CHANGES_REQUESTED** [review]
I think is all good in general, just have a couple of minor comments / questions / checks.

File: aiida/cmdline/commands/cmd_archive.py (hunk 1 above)
Comment (sphuber) [reply aiida/cmdline/commands/cmd_archive.py:440]: I originally added this function (in 65c60d5ef8f3e1169a696fe095322e8163543877) as a shortcut to places in the code that are early in the bootstrapping process and so don't have the `Config` instance loaded yet. @chrisjsewell then refactored it in a8b3da36a0c3d7007519bb563774d38f2ce6b2e2 and moved it to a method on the `Manager`. Think it would make sense to discuss whether there is use to keeping both or we remove one of them. But that would be for another PR.

File: aiida/engine/processes/calcjobs/calcjob.py (hunk 2 above)
Comment (sphuber) [reply aiida/engine/processes/calcjobs/calcjob.py:432]: True, there aren't any guidelines. I think there is something to be said for both, but I also don't think it matters that much. Feel free to open a discussion if you think it would be good to have a guideline for this.

File: tests/cmdline/commands/test_archive_import.py (hunk 4 above)
Comment (sphuber) [reply tests/cmdline/commands/test_archive_import.py:13]: Yeah, this is just a refactoring to use the pytest fixtures. I think that since I removed the class and just made them normal functions, the deindent makes almost every line seem changed. There are no functional changes though, just refactoring.

File: aiida/storage/sqlite_temp/backend.py (hunk 3 above)
Comment (sphuber) [reply aiida/storage/sqlite_temp/backend.py:136]: Good catch, I didn't make this decision consciously, I think I just missed it. Maybe for consistency it would be good to have the same behavior here and also use the option.

File: tests/cmdline/commands/test_archive_import.py (hunk 4 above)
Comment (ramirezfranciscof) [reply tests/cmdline/commands/test_archive_import.py:13]: Ok, I see. In the future maybe it would be good to separate the commits in which we do this kind of refactoring into a different PR, specially if it is not using the feature changed in the PR. I hope it doesn't sound to nagging, its just that it is a bit more mentally costly to be going over this without being clear on the type of changes I'm supposed to be looking for, more so since github is not properly identifying the changes.

Comment (sphuber) [reply tests/cmdline/commands/test_archive_import.py:13]: Good point, it's just that I came across it and it was so tempting to refactor it as it made it so much more succinct. But I agree that it is probably best left to separate refactor PRs.

**Comment by sphuber** [comment]
Cheers @ramirezfranciscof . More reviews would have been great, but it has been open for almost three weeks with plenty of invitations and have had no response, so I will go ahead and merge this.
